core/src/main/java/org/infinispan/reactive/publisher/impl/PublisherHandler.java

```diff
@@ -236,7 +238,19 @@ void startProcessing(InitialPublisherCommand command) {
                   command.isIncludeLoader(), command.getDeliveryGuarantee(), command.getTransformer());
          }
 
-         sap.subscribe(this, this::segmentComplete, this::segmentLost);
+         Flowable.<SegmentAwarePublisher.NotificationWithLost<Object>>fromPublisher(sap::subscribeWithLostSegments)
+               .doOnNext(notification -> {
```

Comment:
Yeah, this is the same as https://github.com/infinispan/infinispan/pull/9608#discussion_r737508072. I will see what I can do.

Reply:
Actually, I am going to remove all of it and just move it into the `onNext` call instead. Seems a lot better.

Reply:
Which includes https://github.com/infinispan/infinispan/pull/9608#discussion_r738745097 in it.

Reply:
I changed it to `mapOptional`, I can't change it to use `onNext` because of the `KeyPublisherState` extending this one. I will have to address this fully in another PR, but this should be a bit cleaner at least.